### DISCUSSION TITLE
Writing libraries and test scope injection problems

### DISCUSSION THREAD

**Discussion by mensinda** [quarkusio/quarkus#25929] [Q&A] [answered]
Hello,

I am writing a set of libraries (with CDI injection) that are each split into an API and (one or more) IMPL module(s). In this example I have two libraries (`A` and `B`) with one API and IMPL module each (`A_api`, `A_impl`, `B_api`, and `B_impl`)

These libraries can depend on each other (here the `B_impl` depends on the `A_api`). Since I want to test `B_impl` (JUnit and `maven verify`), I also have to add a maven dependency to `A_impl`. The problem I have is that `<scope>test</scope>` does not work here since and I get  `Unsatisfied dependency for type...` errors from the `quarkus-maven-plugin`. So, I either have to use the `compile` or the `runtime` scope. Using the `compile` scope is out of question, since doing so would completely negate the point of splitting libraries into different API and IMPL modules.

The problem with the `runtime` scope is that it is transitive. However, I don't want that library `B` dictates which IMPL modules other libraries (or the final app) use. I know that I could use maven dependency exclusion to work around this, but this super unclean (especially with many libraries as is the case for me) and requires knowledge of what IMPL a library uses (which can change over different versions).

Is there a way to tell quarkus to also consider `test` scope deps? Alternatively is there a flag that basically tells the `quarkus-maven-plugin` the following:

> I am writing a library, I know that there are missing dependencies, please ignore this because I know what I am doing.

[a comment by an automated account is not shown]

**Comment by mkouba** [comment]
CC @aloubyansky

**Comment by aloubyansky** [comment, marked as answer]
Unit tests (the `test` goal) should work fine with `<scope>test</scope>`. I suppose you are having problems with the integration tests (the `verify` goal), correct?

If you have your repo publicly available or a sample of it, more details about your use-case would help.

Libraries you are developing aren't Quarkus applications themselves, correct? If so, to be able to test them, you need to create applications using them in the testsuite. One way would be to actually create these kind of applications under `integration-tests` (or something) and run tests against them. Another option could be Maven profiles containing extra dependencies that would have to be enabled to run integration tests.

**Reply by mensinda** [reply]
The main project is closed source so I can't share it here. However, I made a [test project](https://github.com/mensinda/jta-quarkus-suspend-bug/tree/testingStuff) (please ignore the project name and use the `testingStuff` branch) to reproduce it.

**Reply by aloubyansky** [reply]
Unless every lib module is meant to be a standalone Quarkus app there is no reason to `quarkus:build` them. AFAIU the use-case, this should be enough:
````
[aloubyansky@localhost jta-quarkus-suspend-bug]$ git diff
diff --git a/app/pom.xml b/app/pom.xml
index 8ca181f..0c78f51 100644
--- a/app/pom.xml
+++ b/app/pom.xml
@@ -36,4 +36,24 @@
             <scope>runtime</scope>
         </dependency>
     </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>${quarkus.platform.group-id}</groupId>
+                <artifactId>quarkus-maven-plugin</artifactId>
+                <version>${quarkus.platform.version}</version>
+                <extensions>true</extensions>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>build</goal>
+                            <goal>generate-code</goal>
+                            <goal>generate-code-tests</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
 </project>
diff --git a/pom.xml b/pom.xml
index 1b9d598..9da40dd 100644
--- a/pom.xml
+++ b/pom.xml
@@ -52,22 +52,6 @@
 
     <build>
         <plugins>
-            <plugin>
-                <groupId>${quarkus.platform.group-id}</groupId>
-                <artifactId>quarkus-maven-plugin</artifactId>
-                <version>${quarkus.platform.version}</version>
-                <extensions>true</extensions>
-                <executions>
-                    <execution>
-                        <goals>
-                            <goal>build</goal>
-                            <goal>generate-code</goal>
-                            <goal>generate-code-tests</goal>
-                        </goals>
-                    </execution>
-                </executions>
-            </plugin>
-
             <plugin>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>${compiler-plugin.version}</version>
@@ -88,6 +72,25 @@
                     </systemPropertyVariables>
                 </configuration>
             </plugin>
+            <plugin>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <version>${surefire-plugin.version}</version>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                        <configuration>
+                             <systemPropertyVariables>
+                                 <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
+                                 <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
+                                 <maven.home>${maven.home}</maven.home>
+                             </systemPropertyVariables>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
         </plugins>
     </build>
 </project>
````

**Reply by mensinda** [reply]
In a more complex use case, would all tests behave the same without running the `quarkus-maven-plugin` on the lib (JPA/Hibernate, JTA/Narayana, etc.)? What happens with all the build time recorders, bytecode optimization, etc.?

I am fairly new to quarkus, so sorry if the answer is obvious.

**Reply by aloubyansky** [reply]
Sure, no problem. Here what you should know. `quarkus-maven-plugin:build` is used to build **and package** the app (in a JAR or a native image). Your tests actually do not require a **packaged** app. In other words, your tests aren't traditional integration tests. They could work as well with the surefire plugin, AFAICS. `quarkus-junit5` includes a Quarkus JUnit5 extension that will bootstrap the app to run the tests **in memory**. It can happen before the packaging phase.
Quarkus also supports `@QuarkusIntegrationTest` that runs tests against a **packaged** app (in prod mode). You can read more about it here https://quarkus.io/guides/getting-started-testing#quarkus-integration-test

**Reply by aloubyansky** [reply]
The reason Quarkus is able to bootstrap the app for a `@QuarkusTest` but failed to build it with the `quarkus-maven-plugin` in your case is that the application is built in different modes in each case. In case of `@QuarkusTest` the application is built in `test` mode which takes into account `test` scoped dependencies. `quarkus-maven-plugin:build` builds a production app, which won't include test dependencies.

**Reply by mensinda** [reply]
Thanks for the explanation. Just to make sure I have understood it correctly: The `quarkus-maven-plugin` has no effect on unit *and* integration tests (all our tests run during `mvn verify` for reasons) as long as `@QuarkusTest` is used (and not `@QuarkusIntegrationTest`)?

**Reply by aloubyansky** [reply]
Correct.

**Reply by aloubyansky** [reply]
Actually, I take that back :) `quarkus-maven-plugin:build` has no effect on `@QuarkusTest` but `generate-code` and `generate-code-tests` may in case there are gRPC Quarkus extensions that need to generate resources prior to running the tests.
